File: packages/wasm-api-dom/src/dom.ts
import { adaptDPI } from "@thi.ng/adapt-dpi";
import type { NumOrString } from "@thi.ng/api";
import { assert } from "@thi.ng/errors/assert";
import {
	IWasmAPI,
	ReadonlyWasmString,
	WasmBridge,
	WasmStringSlice,
	WasmType,
	WasmTypeBase,
} from "@thi.ng/wasm-api";
import { ObjectIndex } from "@thi.ng/wasm-api/object-index";
import {
	$CreateCanvasOpts,
	$CreateElementOpts,
	$Event,
	$WindowInfo,
	AttribType,
	CreateElementOpts,
	Event as WasmEvent,
	EventBody,
	EventType,
	NS_PREFIXES,
	WasmDomExports,
	WasmDomImports,
} from "./api.js";

/**
 * Hidden property for managed DOM elements to track IDs of attached WASM event
 * listeners
 */
const __listeners = "__wasm_listeners";

interface WasmElement extends Element {
	[__listeners]: Set<number>;
}

/**
 * Map of JS event name regexps to {@link EventType} enums and {@link EventBody}
 * field names
 */
const EVENT_MAP: [
	RegExp,
	Exclude<keyof EventBody, keyof WasmTypeBase> | undefined,
	EventType
][] = [
	[/^drag(end|enter|leave|over|start)|drop$/, "drag", EventType.DRAG],
	[/^blur|focus(in|out)?$/, , EventType.FOCUS],
	[/^change|(before)?input$/, "input", EventType.INPUT],
	[/^key(down|press|up)$/, "key", EventType.KEY],
	[
		/^(dbl)?click|contextmenu|mouse(down|enter|leave|move|out|over|up)$/,
		"mouse",
		EventType.MOUSE,
	],
	[
		/^(got|lost)pointercapture|pointer(cancel|down|enter|leave|move|out|over|up)$/,
		"pointer",
		EventType.POINTER,
	],
	[/^scroll$/, "scroll", EventType.SCROLL],
	[/^touch(cancel|end|move|start)$/, "touch", EventType.TOUCH],
	[/^wheel$/, "touch", EventType.WHEEL],
];

/** @internal */
interface WasmListener {
	ctx: number;
	name: string;
	event: WasmEvent;
	fn: EventListener;
}

export class WasmDom implements IWasmAPI<WasmDomExports> {
	static readonly id = "dom";

	readonly id = WasmDom.id;

	parent!: WasmBridge<WasmDomExports>;
	$Event!: WasmType<WasmEvent>;
	$CreateElementOpts!: WasmType<CreateElementOpts>;

	elements = new ObjectIndex<Element>({ name: "elements" });
	listeners: Record<number, WasmListener> = {};

	protected currEvent: Event | null = null;
	protected currDataTransfer: DataTransfer | null = null;

	async init(parent: WasmBridge<WasmDomExports>) {
		this.parent = parent;
		if (parent.exports._dom_init) {
			parent.exports._dom_init();
		} else {
			parent.logger.warn("DOM module unused, skipping auto-init...");
		}
		this.elements.add(document.head);
		this.elements.add(document.body);
		this.$Event = $Event(this.parent);
		this.$CreateElementOpts = $CreateElementOpts(this.parent);
		return true;
	}

	getImports(): WasmDomImports {
		return {
			getWindowInfo: (ptr: number) => {
				const info = $WindowInfo(this.parent).instance(ptr);
				info.innerWidth = window.innerWidth;
				info.innerHeight = window.innerHeight;
				info.dpr = window.devicePixelRatio || 1;
				info.scrollX = window.scrollX;
				info.scrollY = window.scrollY;
				info.fullscreen =
					(document.fullscreenElement ||
					(<any>document).webkitFullscreenElement
						? 1
						: 0) |
					(document.fullscreenEnabled ||
					(<any>document).webkitFullscreenEnabled
						? 2
						: 0);
			},

			getElementByID: (nameAddr: number) => {
				const name = this.parent.getString(nameAddr);
				let id = this.elements.find((el) => el.id === name);
				if (id === undefined) {
					const el = document.getElementById(name);
					return el ? this.elements.add(el) : -1;
				}
				return id;
			},

			createElement: (optsAddr: number) => {
				const create = (
					opts: CreateElementOpts,
					nestedParent?: number
				) => {
					const tagName = opts.tag.deref();
					const ns = opts.ns.deref();
					const el = ns
						? document.createElementNS(
								NS_PREFIXES[ns] || ns,
								tagName
						  )
						: document.createElement(tagName);
					const id = this.elements.add(el);
					this.initElement(id, el, opts, nestedParent);
					if (opts.children.length > 0) {
						for (let child of opts.children) {
							create(child, id);
						}
					}
					return id;
				};
				return create(this.$CreateElementOpts.instance(optsAddr));
			},

			removeElement: (elementID: number) => {
				assert(elementID > 1, "can't remove reserved element");
				const el = this.elements.get(elementID, false);
				if (!el) return;
				const remove = (el: Element) => {
					const elementID = this.elements.find(
						(x) => x === el,
						false
					);
					if (elementID !== undefined) {
						this.elements.delete(elementID, false);
						const elementListeners = (<WasmElement>el)[__listeners];
						if (elementListeners) {
							for (let listenerID of elementListeners) {
								this.removeListener(el, listenerID);
								// WASM side cleanup
								this.parent.exports._dom_removeListener(
									listenerID
								);
							}
						}
					}
					el.parentNode?.removeChild(el);
					for (let child of [...el.children]) remove(child);
				};
				remove(el);
			},

			createCanvas: (optsAddr: number) => {
				const opts = $CreateCanvasOpts(this.parent).instance(optsAddr);
				const el = document.createElement("canvas");
				adaptDPI(el, opts.width, opts.height, opts.dpr);
				const id = this.elements.add(el);
				this.initElement(id, el, opts);
				return id;
			},

			setCanvasSize: (
				elementID: number,
				width: number,
				height: number,
				dpr: number
			) =>
				adaptDPI(
					<HTMLCanvasElement>this.elements.get(elementID),
					width,
					height,
					dpr
				),

			setStringAttrib: (elementID: number, name: number, val: number) =>
				this.setAttrib(elementID, name, this.parent.getString(val)),

			setNumericAttrib: (elementID: number, name: number, val: number) =>
				this.setAttrib(elementID, name, val),

			_setBooleanAttrib: (
				elementID: number,
				nameAddr: number,
				val: number
			) => {
				const el = this.elements.get(elementID);
				const name = this.parent.getString(nameAddr);
				if (name in el) {
					// @ts-ignore
					el[name] = !!val;
				} else {
					val ? el.setAttribute(name, "") : el.removeAttribute(name);
				}
			},

			_getStringAttrib: (
				elementID: number,
				nameAddr: number,
				valAddr: number,
				maxBytes: number
			) =>
				this.parent.setString(
					String(this.getAttrib(elementID, nameAddr) || ""),
					valAddr,
					maxBytes,
					true
				),

			_getStringAttribAlloc: (
				elementID: number,
				nameAddr: number,
				slice: number
			) =>
				new WasmStringSlice(this.parent, slice).setAlloc(
					String(this.getAttrib(elementID, nameAddr) || ""),
					true
				),

			getNumericAttrib: (elementID: number, nameAddr: number) =>
				Number(this.getAttrib(elementID, nameAddr) || ""),

			_getBooleanAttrib: (elementID: number, nameAddr: number) =>
				~~(this.getAttrib(elementID, nameAddr) != null),

			addClass: (elementID: number, name: number) =>
				this.elements
					.get(elementID)
					.classList.add(this.parent.getString(name)),

			removeClass: (elementID: number, name: number) =>
				this.elements
					.get(elementID)
					.classList.remove(this.parent.getString(name)),

			_addListener: (ctxID: number, name: number, listenerID: number) => {
				const ctx = ctxID < 0 ? window : this.elements.get(ctxID);
				const eventName = this.parent.getString(name);
				const eventSpec = EVENT_MAP.find(([re]) => re.test(eventName));
				const [eventBodyID, eventTypeID] = eventSpec
					? [eventSpec[1], eventSpec[2]]
					: [undefined, EventType.UNKOWN];
				const hasModifiers = [
					EventType.DRAG,
					EventType.INPUT,
					EventType.KEY,
					EventType.MOUSE,
					EventType.POINTER,
					EventType.TOUCH,
					EventType.WHEEL,
				].includes(eventTypeID);
				const event = this.$Event.instance(
					this.parent.allocate(this.$Event.size)[0]
				);
				const body = eventBodyID ? event.body[eventBodyID] : undefined;
				const fn = (e: Event) => {
					this.currEvent = e;
					this.parent.ensureMemory();
					event.__bytes.fill(0);
					const target =
						e.target === ctx
							? ctxID
							: e.target === window
							? -1
							: this.elements.find((x) => x === e.target, false);
					event.target = target !== undefined ? target : -2;
					event.id = eventTypeID;
					const slice = body ? body.fromEvent(<any>e) : undefined;
					if (hasModifiers) {
						(<any>body!).modifiers = this.encodeModifiers(
							<KeyboardEvent>e
						);
					}
					if (eventTypeID === EventType.DRAG) {
						this.currDataTransfer = (<DragEvent>e).dataTransfer;
					}
					this.parent.exports._dom_callListener(
						listenerID,
						event.__base
					);
					if (slice) this.parent.free(slice);
					this.currEvent = null;
					this.currDataTransfer = null;
				};
				this.parent.logger.debug(
					`ctx ${ctxID} - adding ${eventName} listener #${listenerID}`
				);
				ctx.addEventListener(eventName, fn);
				this.listeners[listenerID] = {
					ctx: ctxID,
					name: eventName,
					event,
					fn,
				};
				if (ctxID >= 0) {
					(
						(<WasmElement>ctx)[__listeners] ||
						((<WasmElement>ctx)[__listeners] = new Set())
					).add(listenerID);
				}
			},

			preventDefault: () => {
				this.currEvent && this.currEvent.preventDefault();
			},

			stopPropagation: () => {
				this.currEvent && this.currEvent.stopPropagation();
			},

			stopImmediatePropagation: () => {
				this.currEvent && this.currEvent.stopImmediatePropagation();
			},

			_removeListener: (listenerID: number) => {
				const listener = this.listeners[listenerID];
				assert(!!listener, `unknown listener ID: ${listenerID}`);
				const ctx =
					listener.ctx < 0 ? window : this.elements.get(listener.ctx);
				this.removeListener(ctx, listenerID);
				if (listener.ctx >= 0) {
					const listeners = (<WasmElement>ctx)[__listeners];
					if (listeners.has(listenerID)) listeners.delete(listenerID);
				}
			},

			setInnerHtml: (elementID: number, body: number) => {
				this.elements.get(elementID).innerHTML =
					this.parent.getString(body);
			},

			setInnerText: (elementID: number, body: number) => {
				(<HTMLElement>this.elements.get(elementID)).innerText =
					this.parent.getString(body);
			},

			_requestAnimationFrame: (rafID: number) => {
				this.parent.logger.fine(`requestAnimationFrame #${rafID}`);
				requestAnimationFrame((t) =>
					this.parent.exports._dom_callRAF(rafID, t)
				);
			},

			_requestFullscreen: async (elementID: number) => {
				if (
					!(
						document.fullscreenElement ||
						(<any>document).webkitFullscreenElement
					)
				) {
					const el =
						elementID <= 1
							? document.documentElement
							: this.elements.get(elementID);
					const method =
						el.requestFullscreen ||
						(<any>el).webkitRequestFullscreen;
					await method.bind(el)();
					this.parent.exports._dom_fullscreenChanged();
				}
			},

			_exitFullscreen: async () => {
				if (
					document.fullscreenElement ||
					(<any>document).webkitFullscreenElement
				) {
					const method =
						document.exitFullscreen ||
						(<any>document).webkitExitFullscreen;
					await method.bind(document)();
					this.parent.exports._dom_fullscreenChanged();
				}
			},
		};
	}

	protected initElement(
		elementID: number,
		el: Element,
		opts: Pick<
			Readonly<CreateElementOpts>,
			"attribs" | "class" | "id" | "index" | "parent"
		> &
			Partial<{
				html: ReadonlyWasmString;
				text: ReadonlyWasmString;
			}>,
		nestedParent?: number
	) {
		const { id, attribs, class: $class, index } = opts;
		if (id.length) el.setAttribute("id", id.deref());
		if ($class.length) el.setAttribute("class", $class.deref());
		if (opts.html?.length) {
			el.innerHTML = opts.html.deref();
		} else if (opts.text?.length) {
			(<HTMLElement>el).innerText = opts.text.deref();
		}
		if (attribs && attribs.length) {
			for (let attr of attribs) {
				const name = attr.name.deref();
				if (attr.kind === AttribType.EVENT) {
					const listenerAddr = attr.value.event.__base;
					const listenerID =
						this.parent.exports._dom_addListener(listenerAddr);
					this.getImports()._addListener(
						elementID,
						attr.name.addr,
						listenerID
					);
				} else if (attr.kind === AttribType.FLAG) {
					attr.value.flag && el.setAttribute(name, "");
				} else {
					el.setAttribute(
						name,
						attr.kind === AttribType.STR
							? attr.value.str.deref()
							: String(attr.value.num)
					);
				}
			}
		}
		const parent = nestedParent != undefined ? nestedParent : opts.parent;
		if (parent >= 0) {
			const parentEl = this.elements.get(parent);
			index < 0
				? parentEl.appendChild(el)
				: parentEl.insertBefore(el, parentEl.childNodes[index]);
		}
	}

	protected encodeModifiers(e: KeyboardEvent) {
		return (
			(e.shiftKey ? 1 : 0) |
			(e.ctrlKey ? 2 : 0) |
			(e.altKey ? 4 : 0) |
			(e.metaKey ? 8 : 0)
		);
	}

	protected getAttrib(elementID: number, nameAddr: number) {
		const el = this.elements.get(elementID);
		const name = this.parent.getString(nameAddr);
		return name in el ? el[<keyof Element>name] : el.getAttribute(name);
	}

	protected setAttrib(
		elementID: number,
		nameAddr: number,
		value: NumOrString
	) {
		const el = this.elements.get(elementID);
		const name = this.parent.getString(nameAddr);
		return name in el
			? // @ts-ignore
			  (el[name] = value)
			: el.setAttribute(name, String(value));
	}

	protected removeListener(ctx: Window | Element, listenerID: number) {
		const listener = this.listeners[listenerID];
		assert(!!listener, `invalid listener ID ${listenerID}`);
		this.parent.logger.debug(`removing event listener #${listenerID}`);
		delete this.listeners[listenerID];
		ctx.removeEventListener(listener.name, listener.fn);
		this.parent.free([listener.event.__base, this.$Event.size]);
	}
}

File: packages/wasm-api-dom/zig/dom.zig
const std = @import("std");
const dom = @import("api.zig");
const ManagedIndex = @import("wasmapi").ManagedIndex;

pub usingnamespace dom;

/// DOM event listener
pub const EventListener = struct {
    /// Event listener function. Takes an event and optional pointer to user
    /// supplied arbitrary context data provided when registering the handler
    /// via addListener()
    callback: *const fn (event: *const dom.Event, ?*anyopaque) void,
    /// Optional type erased pointer to arbitrary user context. This pointer
    /// can be cast back into the desired type using this form:
    /// `@ptrCast(?*Foo, @alignCast(@alignOf(Foo), raw))`
    /// Also see: `wasmapi.ptrCast()`
    ctx: ?*anyopaque = null,
};

/// RAF event listener
pub const RAFListener = struct {
    /// Animation frame listener function. Takes an hires timestamp and optional
    /// pointer to user supplied arbitrary context data provided when registering
    /// the handler via requestAnimationFrame()
    callback: *const fn (time: f64, ?*anyopaque) void,
    /// Optional type erased pointer to arbitrary user context. This pointer
    /// can be cast back into the desired type using this form:
    /// `@ptrCast(?*Foo, @alignCast(@alignOf(Foo), raw))`
    /// Also see: `wasmapi.ptrCast()`
    ctx: ?*anyopaque = null,
};

pub const FullscreenCallback = *const fn () void;

/// Reserved reference handle for the browser window itself (e.g. used for event targets)
pub const WINDOW: i32 = -1;
/// Reserved reference handle for `document.body`
pub const DOC_BODY: i32 = 0;

pub const DOMError = error{
    InvalidID,
};

var eventListeners: ManagedIndex(*const EventListener, u16) = undefined;
var rafListeners: ManagedIndex(*const RAFListener, u16) = undefined;

pub fn init(allocator: std.mem.Allocator) !void {
    eventListeners = ManagedIndex(*const EventListener, u16).init(allocator);
    rafListeners = ManagedIndex(*const RAFListener, u16).init(allocator);
}

pub extern "dom" fn getWindowInfo(desc: *dom.WindowInfo) void;

pub extern "dom" fn createElement(opts: *const dom.CreateElementOpts) i32;

pub extern "dom" fn removeElement(elementID: i32) bool;

pub extern "dom" fn createCanvas(opts: *const dom.CreateCanvasOpts) i32;

pub extern "dom" fn setCanvasSize(elementID: i32, width: u16, height: u16, dpr: u8) void;

pub extern "dom" fn _setStringAttrib(elementID: i32, name: [*]const u8, val: [*]const u8) void;

/// Sets attrib for given name to string val
pub fn setStringAttrib(elementID: i32, name: []const u8, val: []const u8) void {
    _setStringAttrib(elementID, name.ptr, val.ptr);
}

pub extern "dom" fn _getStringAttrib(elementID: i32, name: [*]const u8, val: [*]u8, maxBytes: usize) usize;

/// Returns string value of attrib for given name
pub fn getStringAttrib(elementID: i32, name: []const u8, val: []u8) []u8 {
    return val[0.._getStringAttrib(elementID, name.ptr, val.ptr, val.len)];
}

pub extern "dom" fn _setNumericAttrib(elementID: i32, name: [*]const u8, val: f64) void;

/// Sets attrib for given name to numeric val
pub fn setNumericAttrib(elementID: i32, name: []const u8, val: f64) void {
    _setNumericAttrib(elementID, name.ptr, val);
}

pub extern "dom" fn _getNumericAttrib(elementID: i32, name: [*]const u8) f64;

/// Returns numeric value of attrib for given name
pub fn getNumericAttrib(elementID: i32, name: []const u8) f64 {
    return _getNumericAttrib(elementID, name.ptr);
}

pub extern "dom" fn _setInnerHtml(elementID: i32, ptr: [*]const u8) void;

/// Sets the `.innerHTML` property of a DOM element to given string
pub fn setInnerHtml(elementID: i32, html: []const u8) void {
    _setInnerHtml(elementID, html.ptr);
}

pub extern "dom" fn _setInnerText(elementID: i32, ptr: [*]const u8) void;

/// Sets the `.innerText` property of a DOM element to given string
pub fn setInnerText(elementID: i32, text: []const u8) void {
    _setInnerText(elementID, text.ptr);
}

/// Internal callback. Called from JS
export fn dom_callListener(listenerID: u16, event: *const dom.Event) void {
    if (eventListeners.get(listenerID)) |listener| listener.callback(event, listener.ctx);
}

pub extern "dom" fn _addListener(elementID: i32, name: [*]const u8, listenerID: u16) void;

/// Adds given listener to a DOM element for event `name`.
/// Returns an unique listener ID.
pub fn addListener(elementID: i32, name: []const u8, listener: *const EventListener) !u16 {
    const listenerID = try eventListeners.add(listener);
    _addListener(elementID, name.ptr, listenerID);
    return listenerID;
}

pub extern "dom" fn _removeListener(listenerID: u16) void;

/// Removes the listener for given ID
pub fn removeListener(listenerID: u16) void {
    eventListeners.remove(listenerID);
    _removeListener(listenerID);
}

/// Calls .preventDefault() on currently processed event
/// (only to be called from an EventListener!)
pub extern "dom" fn preventDefault() void;

/// Calls .stopPropagation() on currently processed event
/// (only to be called from an EventListener!)
pub extern "dom" fn stopPropagation() void;

/// Calls .stopImmediatePropagation() on currently processed event
/// (only to be called from an EventListener!)
pub extern "dom" fn stopImmediatePropagation() void;

pub extern "dom" fn _requestAnimationFrame(listenerID: u16) void;

/// Registers given listener for next animation frame
pub fn requestAnimationFrame(listener: *const RAFListener) !u16 {
    const id = try rafListeners.add(listener);
    _requestAnimationFrame(id);
    return id;
}

/// Internal callback. Called from JS
export fn dom_callRAF(listenerID: u16, time: f64) void {
    if (rafListeners.get(listenerID)) |raf| {
        rafListeners.remove(listenerID);
        raf.callback(time, raf.ctx);
    }
}

var fsCallback: ?FullscreenCallback = null;

pub extern "dom" fn _requestFullscreen(elementID: i32) void;

pub extern "dom" fn _exitFullscreen() void;

/// Only to be called from an event handler. Requests fullscreen display for
/// given element ID. Use -1 for the entire window/document.
/// The optional callback will be called once (and if) fullscreen is enabled
pub fn requestFullscreen(elementID: i32, callback: ?FullscreenCallback) void {
    fsCallback = callback;
    _requestFullscreen(elementID);
}

/// Only to be called from an event handler. Exits fullscreen display (if
/// active). The optional callback will be called once (and if) fullscreen
/// switch is complete.
pub fn exitFullscreen(callback: ?FullscreenCallback) void {
    fsCallback = callback;
    _exitFullscreen();
}

/// Internal callback. Called from JS
export fn dom_fullscreenChanged() void {
    if (fsCallback) |callback| {
        callback();
        fsCallback = null;
    }
}
